### DISCUSSION TITLE
Testing with global context-provided code

### DISCUSSION THREAD

**Discussion by ohad-griiip** [pytest-dev/pytest#9330] [Q&A] [answered]
My tests structure looks like this - 
```
tests/
    __init__.py
    test_no_mocked_library.py
    tests_with_mocked_library/
        conftest.py
        ....
```
In the nested _conftest_ I use the _pytest_collectstart_ like this - 
```py
def pytest_collectstart(collector):
    os.environ['API_KEY'] = '...'
    sys.modules['somelib'] = Mock()
```
Without this the tests wouldn't run (the code there're importing would raise).
Using this hook I manage to run those tests but then the mocked library remains mocked also in _test_no_mocked_library_ which causes a different problem. 

I didn't find an opposite hook to _pytest_collectstart_ to use for cleanup. 
Is there a way to be able to make all those tests pass together?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
I would strongly recommend a scoped fixture, hacks with sys. Modules make for fragile tests, and I typically strongly recommended against them

**Reply by ohad-griiip** [reply]
Thanks for replying @RonnyPfannschmidt! I totally agree that mocking modules and over use of mock and patching is bad and fragile. 
This specific code is a bit messy, but still I would like to get your recommendation on something like this (more standard I'd say) - 

Let's say I have a configuration - 
```py
# settings.py

class Settings(BaseSettings):
    ...
    
 def get_config() -> Settings:
    ...
``` 
That would load from env, file, db, api, ... 
Plus I have an api module - `api.py` that wraps requests. 

How would you test this api without patching/mocking, should it involve some dependency injection of api object in the settings or there is a different approach? 
And about setting test configuration, what do you think about using the _pytest_collectstart_ to set env variables so when the tests run and imports the source code the configuration can be loaded - how would you have different tests testing different configuration without patching?

**Reply by RonnyPfannschmidt** [reply]
for projects im involved in i typically want to completely avoid global configuration objects, over the years those have been the reason for serious grief in multiple ways (a number of projects that started out simple and then had to add different configruations in the same test we ended up with serious issues about isolation of the configuration (in one occasion a  destructive test accidentally ran against a stage environment instead of its designated environment as config switch-over was fragile and there was a unexpected cache)

i strongly advise to always create configured application objects that get the configuration passed in as dependency to avoid those horrors

**Reply by ohad-griiip** [reply]
So are you against patterns of _provider_ even in cases of chain of handing parameters from object/module to another? 
It may be beneficial for tests (for isolation etc) but can make the code ugly and leggy no? 
What about listening/registering things like global events do you pass each such function an object specifically or can it import some function from a global "store"?

I do face issues with some code designs I have trying to get insights on better patterns. 
On one hand I want to have good testable code but I wouldn't want it to be too tedious... 🤔

**Reply by RonnyPfannschmidt** [reply]
To excessively over paint the picture :

There are no nice patterns for inspection of Spagetti when it's finally on the plate.

Testability as an afterthought is akin to trying to arrange a completed spaghetti dish back to how the uncooked Spagetti arrived

It's not clear what you mean by "chain" without an example i can't make a sensibel point about it

**Reply by RonnyPfannschmidt** [reply]
As for provider, isn't that just a fixture?

**Reply by RonnyPfannschmidt** [reply]
As for global events, that's not a concept that can exist in a system where there is no actual global configuration

Events would register against instances

**Reply by ohad-griiip** [reply]
@RonnyPfannschmidt 100% agree with the spaghetti part, I think I should change the title to "testing provider pattern" now 😄.  
In _provider_ pattern I mean patterns when one part of the code can get things from context, a global context in my case.
For example the configuration I've mentioned above with the `get_config` function in some _settings_ module, a module/function elsewhere in the code can do something like - 

```py
def handle_something():
    cfg = get_config()
    wait_on_something(timeout=cfg.timeout)
``` 

It this case the function doesn't need a _timeout_ to be explicitly passed. 
Or registering to something - 
```py
from events import register_handler

@register_handler('...')
def on_change():
   ...
```

I find this pattern to save boilerplate (when due) and having to pass parameters from hand to hand.
Logging is another example for that when we use `logging.getLogger` instead of passing logger directly to modules/class/functions. 

👉 Now when it comes to testability it seems easier if everything is passed directly to a function (e.g a logger), removing the need for mock.
I wonder [a] what do you think about provider pattern and [b] good way to do testing.

**Reply by ohad-griiip** [reply]
BTW, though playing with sys modules may not be good practice I thought maybe I could do something like - 
```py
# conftest.py

def pytest_generate_tests():
   for module in {module in sys.modules if module.startswith('mypackage')}:
       sys.modules.pop(module)
```
So python would reimport the package before each test thus getting the "context" created again each time. 
Didn't try it yet perhaps it can slow down the process, though I may ended up with sub tests packages each one setting up the configuration of the code (via env variables or something like this) for its tests and do this modules popping only before the collecting phase.

**Reply by RonnyPfannschmidt** [reply]
in general its dangerous and error prone to remove items from `sys.modules`

as for the `handle_something` function - code like that is just a time bomb - if not managed well, at some point you will just confidentially run destructive code against the wrong configuration

as for not explicitly passing things,
objects and methods exist for a reason,

as for "globalized events" sqlalchemy has prime examples of making it sensible 

```python

from .config import MyConfig
import pytest

class Myhandler:
   config: MyConfig
   
   def handle_something(self):
      wait_for_something(timeout=self.config.timeout)

def get_config_for_test(request: pytest.FixtureRequest):
   return MyConfig(...)
      
@fixture
def my_handler(request: pytest.FixtureRequest) ->  MyHanbdler:
   return MyHandler(get_config_for_test(request))
   
def test_something(my_handler: MyHandler):
  my_handler.handle_something()
  ```

**Reply by ohad-griiip** [reply]
Ok let me share a use case I have - 
I have a reactive configuration meaning that one can subscribe for changes something like this - 

```py
# settings.py

class Config:
    x: ...
    y: ...

# module.py

cfg = get_config()

class Device:
    def __init__(self):
        ...

    @cfg.subscribe('x')
    def handle_x_change(prev, current):
        ...
```
The reactivity is something I need and I thought this could be a good/declarative implementation.
If I understand you correctly you would go for handing a configuration object to `Device.__init__` instead of calling the _get_config_ and then I could register callbacks without the decorator style, correct? 

But then what happens code-wise when this piece calls another one and passes it the configuration (of some sub fields of it)...
If I have an api module for example which is also kind of global should it be a class that gets a config object instead of functions and then other classes/functions should be handed this api?  
Perhaps it's not too bad and I just need to accept and get used to it, if you say this pattern is better I'll take your word on it.

**Reply by RonnyPfannschmidt** [reply]
It can be turned declarative without needing globals, but then one needs to actually have a layer in place that takes the declaration and runs the registration

An example of such a toolkit is the dectate library, however it is not widely used

It can be error prone to have a process /app continuously converge after config changes,
My personal preference is to start new processes, that however may not fit the use case with devices you outline

**Reply by ohad-griiip** [reply]
I took I little time to think about it and got into the conclusion that the code I have is simply unrightfully nested and needs to be organized and simplified. I understand why I should avoid touching sys.modules and having global objects that can make it harder to have isolated tests. 
After thinking I think that probably the cases where one truly can pick up provider and context hierarchy is when having really nested objects like UI, but not in other scenarios. 

Thanks for the insights @RonnyPfannschmidt!
